lib/src/model/broadcast/broadcast_game_controller.dart

```diff
@@ -0,0 +1,112 @@
+import 'dart:async';
+
+import 'package:deep_pick/deep_pick.dart';
+import 'package:lichess_mobile/src/model/analysis/analysis_controller.dart';
+import 'package:lichess_mobile/src/model/broadcast/broadcast_providers.dart';
+import 'package:lichess_mobile/src/model/common/chess.dart';
+import 'package:lichess_mobile/src/model/common/id.dart';
+import 'package:lichess_mobile/src/model/common/socket.dart';
+import 'package:lichess_mobile/src/network/socket.dart';
+import 'package:lichess_mobile/src/utils/json.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'broadcast_game_controller.g.dart';
+
+@riverpod
+class BroadcastGameController extends _$BroadcastGameController {
+  static Uri broadcastSocketUri(BroadcastRoundId broadcastRoundId) =>
+      Uri(path: 'study/$broadcastRoundId/socket/v6');
+
+  StreamSubscription<SocketEvent>? _subscription;
+
+  late SocketClient _socketClient;
+
+  @override
+  Future<String> build(BroadcastRoundId roundId, BroadcastGameId gameId) async {
```

Comment:
The signature of this controller indicates something is wrong. 

It is returning a PGN that never changes. A riverpod notifier that does not update its state should not exist really. See the comment below where I explain in more details what should be done instead.

 Under the hood it is imperatively changing the analysis controller state. That is another pattern that we should avoid with riverpod. This is a reactive framework so instead of:

```
Provider A
----------
subs to an event source
on each event --> modify Provider B
```

we should do:
```
Provider A
----------
subs to an event source
on each event: --> update and notifies its state change

Provider B
----------
subs to provider A changes
on each change: do whatever needed
```

A provider modifying another one imperatively is almost never what we want. I've done it myself a couple of times too and realised later it was a mistake. It is possible there is still a couple of bad patterns like that in the app code but it should be fixed really.

Reply:
I didn't even realize that I was not updating the state of this controller. I think this is due the way I was reasoning. My goal was to update the analysis controller from moves received by a broadcast like it is for moves received by user inputs.

I am not sure about changing the signature to `Future<BroadcastAnalysisState>`. If it is a future it will probably make it harder to implement a mixin for both controller (`BroadcastGameController` and `AnalysisController`)

Reply:
The `StudyController` from https://github.com/lichess-org/mobile/pull/1111 already returns Future<StudyState> as well, so you won't make refactoring harder

Reply:
Indeed `StudyController` and `BroadcastController` will be very similar. We'll be able to refactor all this step by step. 

Starting with a mixin for engine evaluation I think, as it looks straightforward.

Reply:
I created a controller that is separated of the analysis controller but I still believe that is not the right way to do it. Now the code is full of `if (!state.hasValue) return;` and `state.requireValue` whereas the state will never be a future expect during the loading of the pgn which is only when a user taps on a game. 

Also the factoring of `BroadcastController` methods and `AnalysisController` methods will be harder since the methods are not the same anymore because of the check for a value in the state (`if (!state.hasValue) return;` and `state.requireValue` )

Reply:
Yes I know it's annoying, having to deal with `AsyncValue`. But actually there is no other way, is it?

How do you deal with the absence of value (while it is loading) otherwise?

We'll have to make the `AnalysisController` async too.

In order to make this refactoring, we'd have a `GameAnalysisController` which would returns a Future<GameAnalysisState> and takes a game id + options. And a `StandaloneAnalysisController` which takes a PGN string (but has to be async too to be compatible with the common interface).

I don't have a better idea for now. What do you think @tom-anders ? Since you've worked on studies you might have an idea on how to deal with this?

Reply:
So for studies, `AsyncValue` actually feels really natural - For broadcast, the `state` is `loading` only at the very start, but for studies it's actually also `loading` when switching between chapters.

I already did some experimentation locally to factor out common code between `StudyController` and `AnalysisController`, and while it does somehow work, the fact that one had `State` while the other had `AsyncValue<State>` was really annoying, meant I couldn't reuse as much code as I wanted to, and also made readability worse in some places. 

I get that making `AnalysisController` async (even though it technically does not need to be) feels weird, but I think the benefits of improving code reusability would outweigh the benefits here. I already caused a bug in `StudyController` because I incorrectly ported a change I made in the `AnalysisController` on another branch. And yeah, there will be a lot of `state.requireValue` (the `StudyController` has them as well), but at least they will be in a single place.

Reply:
So the plan is to merge study and broadcast branches as-is (with separated code). Then we'll refactor study, broadcast and analysis together, making everything async, and write a lot of tests while doing it. cc @julien4215

Reply:
Sounds good to me, let me know if you need help :)

Reply:
What I could do to make the `broadcastAnalysisController` not async is giving it directly the pgn instead of the broadcastGameId like it is done for the `AnalysisController`. I would get the pgn in a loading screen and then give it to broadcast game controller. Would it be hard to do the same for study ?

Reply:
Not sure if that would work:

1) For studies there's the additional complexity of switching between chapters
2) I think the major problem is that for studies, the PGN does not actually contain all the information that the `StudyController` needs (for example the hints in interactive studies, or whether analysis is allowed or not, etc..), so it would also need to be passed the `Study` object - but IIRC passing such heavy weight objects to a provider family is considered an antipattern in Riverpod

Reply:
On the other hand, if we're making the `AnalysisController` async anyway, then we **could** actually give it the `id` instead of `pgn` and let it internally fetch the game itself from the server — Then it also wouldn't feel as pointless to slap a `Future` around its state

Reply:
Is it possible to do that ? Maybe there are use cases where we can't use an `id` but only a `pgn`.

Reply:
Hmm, the `AnalysisScreen` already uses a `String pgnOrId` field so it should be possible - But separating `GameAnalysisController` and `StandaloneAnalysisController` as veloce suggested makes sense as well  I  guess

Reply:
Yes we need to make everything async. And so a GameAnalysisController that takes no PGN but the GameId makes more sense.

Reply:
So the plan is to make two separate base class : a `GameAnalysisController` and a `StandaloneAnalysisController` ?

The `BroadcastGameController` requires two ids : the round id is to listen to the websocket and the game id to get the pgn.